polyphemus/lib/client/jsx/etl/run-pane.tsx

```diff
@@ -0,0 +1,125 @@
+import React, { useState, useCallback } from 'react';
+import Button from '@material-ui/core/Button';
+import Grid from '@material-ui/core/Grid';
+import MenuItem from '@material-ui/core/MenuItem';
+import Select from '@material-ui/core/Select';
+import Switch from '@material-ui/core/Switch';
+import TextField from '@material-ui/core/TextField';
+import Typography from '@material-ui/core/Typography';
+import InputAdornment from '@material-ui/core/InputAdornment';
+import {makeStyles} from '@material-ui/core/styles';
+import EtlPane, {EtlPaneHeader} from './etl-pane';
+
+import { RUN_ONCE, RUN_NEVER, RUN_INTERVAL, getRunState, getRunIntervalTime, runTime } from './run-state';
+
+const useStyles = makeStyles( theme => ({
+  runbar: {
+    width: 'auto',
+    justifyContent: 'space-between'
+  },
+  title: {
+    flex: '0 0 200px',
+  },
+  values: {
+    flex: '1 1 auto'
+  },
+  params: {
+    flex: '1 1 auto'
+  },
+  param: {
+    borderBottom: '1px solid #eee'
+  }
+}));
+
+const Param = ({value, opts, name, update}:{
+  value: undefined|string|boolean,
+  opts: string[]|'string'|'boolean',
+  name: string,
+  update: (name:string,value:string|boolean) => void
+}) => {
+  if (Array.isArray(opts)) {
+    return <Select value={value||''} onChange={e => update(name, e.target.value as string)}>
+      {
+        opts.map(
+          opt => <MenuItem key={opt} value={opt}>{opt}</MenuItem>
+        )
+      }
+    </Select>;
+  }
+
+  if (opts == 'string') {
+    return <TextField size='small' fullWidth value={value== undefined ? '' : value} onChange={e => update(name, e.target.value)}/>;
+  }
+
+  if (opts == 'boolean') {
+    return <Switch size='small' checked={value == undefined ? false : value as boolean} onChange={ e => update(name, e.target.checked)}/>;
+  }
+
+  return null
+}
+
+const RunPane = ({run_interval,update,params,param_opts,selected}:{run_interval:number, params:any, param_opts: any, update:Function,selected:string|null}) => {
+  const [ runState, setRunState ] = useState(getRunState(run_interval));
+  const [ runIntervalTime, setRunIntervalTime ] = useState(getRunIntervalTime(run_interval));
+  const [ newParams, setParams ] = useState(params);
+
+  const runValue = () => runState == RUN_INTERVAL ? runIntervalTime : runState;
+  const reset = () => { setRunState(getRunState(run_interval)); setRunIntervalTime(getRunIntervalTime(run_interval)); setParams(params) };
+
+  const classes = useStyles();
+
+  return <EtlPane mode='run' selected={selected}>
+    <EtlPaneHeader title='Run'>
+      <Grid className={classes.runbar} spacing={1} container>
+      {
+        (newParams != params || (runState != run_interval && (runState != RUN_INTERVAL || runIntervalTime != run_interval))) && <React.Fragment>
+          <Grid item><Button onClick={ () => update({run_interval: runValue(), params: newParams}) }>Save</Button></Grid>
```

Comment:
This posts an update, which ought to reset the state (and thus make the save button disappear), but I can't quite make it work if you update params, probably because of the weird way I am initializing newParams.

Reply:
If you want `newParams` to be set to the `params` prop, perhaps you can initialize it to some empty value and then use `useEffect` to initialize it when params changes (which sounds like it might be what you want??). Something like:

```
const [newParams, setNewParams] = useState({});

useEffect(() => {
  setNewParams(params);
}, [params]);
```

It's probably not resetting because state won't re-initialize when `params` changes, it's not like a controlled input value.